### PR TITLE
Add support for `archive_existing_versions` [Backend API]

### PR DESCRIPTION
**Pull request by harupy**
## What changes are proposed in this pull request?

- Add support for  `archive_existing_versions` in `transition_model_version_stage`.

## How is this patch tested?

Unit tests

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

`transition_model_verstion_stage` now supports the `archive_existing_versions` flag which allows archiving all existing models in the stage.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [x] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/model_registry/sqlalchemy_store.py [hunk 1]
```diff
@@ -446,19 +446,28 @@ def transition_model_version_stage(self, name, version, stage,
...
+        is_active_stage = get_canonical_stage(stage) in DEFAULT_STAGES_FOR_GET_LATEST_VERSIONS
+        if archive_existing_versions and not is_active_stage:
+            msg_tpl = ("Model version transition cannot archive existing model versions "
+                       "because '{}' is not an Active stage. Valid stages are {}")
+            raise MlflowException(msg_tpl.format(stage, DEFAULT_STAGES_FOR_GET_LATEST_VERSIONS))
+
         with self.ManagedSessionMaker() as session:
+            last_updated_time = now()
+            in_same_stage = SqlModelVersion.current_stage == stage
+            model_versions = session.query(SqlModelVersion).filter(in_same_stage).all()
```
Comment (andychow-db) [inline mlflow/store/model_registry/sqlalchemy_store.py:458]: Is this scoped to a single registered model?

File: tests/store/model_registry/test_sqlalchemy_store.py [hunk 2]
```diff
@@ -352,6 +352,46 @@ def test_update_model_version(self):
...
             self.assertEqual(mvd5.current_stage, "Staging")
 
+    def test_transition_model_version_stage(self):
+        msg = (r"Model version transition cannot archive existing model versions "
+               r"because .+ is not an Active stage. Valid stages are .+")
+
+        for stage in ["None", "Archived"]:
+            with self.assertRaisesRegex(MlflowException, msg):
+                self.store.transition_model_version_stage("model", "1", "None", True)
+
```
Comment (sueann) [inline tests/store/model_registry/test_sqlalchemy_store.py:362]: check that with `False` we don't throw?

File: tests/store/model_registry/test_sqlalchemy_store.py [hunk 3]
```diff
@@ -352,6 +352,46 @@ def test_update_model_version(self):
...
+        name = "model"
+        self._rm_maker(name)
+        mv1 = self._mv_maker(name)
+        mv2 = self._mv_maker(name)
+        mv3 = self._mv_maker(name)
+
```
Comment (sueann) [inline tests/store/model_registry/test_sqlalchemy_store.py:368]: can we also check somewhere that with `False` we don't archive the existing versions?

File: mlflow/store/model_registry/sqlalchemy_store.py (hunk 1 above)
Comment (sueann) [reply mlflow/store/model_registry/sqlalchemy_store.py:458]: mm might want to add a test that other registered model versions are not affected

File: tests/store/model_registry/test_sqlalchemy_store.py (hunk 3 above)
Comment (harupy) [reply tests/store/model_registry/test_sqlalchemy_store.py:368]: I'll add that test

File: tests/store/model_registry/test_sqlalchemy_store.py (hunk 2 above)
Comment (harupy) [reply tests/store/model_registry/test_sqlalchemy_store.py:362]: Makes sense. I'll add it

File: mlflow/store/model_registry/sqlalchemy_store.py (hunk 1 above)
Comment (harupy) [reply mlflow/store/model_registry/sqlalchemy_store.py]: Thanks for pointing this out. I fixed the conditin as follows:

```python
conditions = [
    SqlModelVersion.name == name,
    SqlModelVersion.version != version,
    SqlModelVersion.current_stage == stage,
]
model_versions = session.query(SqlModelVersion).filter(in_same_stage).all()
```

**Review by andychow-db — APPROVED** [review]
LGTM

File: tests/store/model_registry/test_sqlalchemy_store.py [hunk 4]
```diff
@@ -352,20 +352,56 @@ def test_update_model_version(self):
...
         name = "model"
         self._rm_maker(name)
         mv1 = self._mv_maker(name)
         mv2 = self._mv_maker(name)
         mv3 = self._mv_maker(name)
 
+        msg = (r"Model version transition cannot archive existing model versions "
+               r"because .+ is not an Active stage. Valid stages are .+")
+
+        # test that when `archive_existing_versions` is False, transitioning a model version
```
Comment (sueann) [inline tests/store/model_registry/test_sqlalchemy_store.py:399]:
```suggestion
        # test that when `archive_existing_versions` is True, transitioning a model version
```

Comment (harupy) [reply tests/store/model_registry/test_sqlalchemy_store.py]: @sueann Thanks!

**Review by sueann — APPROVED** [review]
lgtm!

**Pull request merged**